README.MD
# Minimal PHP framwork
[![Build Status](https://travis-ci.org/GaetanRoger/minimal-php-framework.svg?branch=master)](https://travis-ci.org/GaetanRoger/minimal-php-framework)
[![Coverage Status](https://coveralls.io/repos/github/GaetanRoger/minimal-php-framework/badge.svg?branch=master)](https://coveralls.io/github/GaetanRoger/minimal-php-framework?branch=master)


## Concept
### General idea
This minimal framework is meant to be used for a Model & Manager PHP based application.

- Model: class that represents one key data object of your application
- Manager: singleton class managing all instances of one model

If your application is about cars, you could have the following classes and methods:
```
- Car
  - drive()
  - stop()
  - getDriver()
- CarManager
  - createNewCar()
  - getCarWithId(int)
```

### With the framework
This minimal framework defines two classes for you: `AbstractModel` and `AbstractManager`.
Both predefine methods that every model and manager need, like `Manager.find(int)`, 
`Manager.findAll()`, `Model.getId()`, `Model.insert()`, etc. (please find all defined methods in [Methods list](#methods-list) down below).

This framework makes data persistence easy by using properties and class names to match database table names.
For example, if a model is named `AwesomeCar` the framework will look for a table named `awesome_car`. 



## How to install
This project is available as a [Composer dependency](https://packagist.org/packages/gaetanroger/minimal-php-framework).
Please use the following command:
```
composer require gaetanroger/minimal-php-framework
```

## How to use
At the beginning of your application, please set the PDO connection in the model and manager like this:
```php
AbstractModel::$database = $yourPDOConnection;
AbstractManager::$database = $yourPDOConnection;
```
This will allow the model to update and insert itself into the database, and the manager to find models (from an ID for 
example). Plus, all your models and manager will be able to easily access database.

Then simply make all your models and managers inherit from `Gaetanroger\MinimalPhpFramework\AbstractModel` and 
`Gaetanroger\MinimalPhpFramework\AbstractManager.` Be careful to follow naming conventions specified down below, as names are by default
dynamically used by the framework.

## Conventions to follow
##### Model
* A model class name **must** use CamelCase;
* A model class name **should not** end with `Model` (except if it is part of the actual model name);
* The model class name **must** be the exact conversion from snake_case to CamelCase of the table name.

Valid model names with their table name:
```
FooBar --> foo_bar
Person --> person
AwesomeMechanicalPiece --> awesome_mechanical_piece
```
Invalid model names:
```
Foo_bar
person
AweSomeMechanicalPiece
```

##### Manager
* A manager class name **must** use CamelCase;
* A manager class name **must** end with `Manager`;
* A manager class name **must** not contain `Manager` anywhere else;
* The manager class name **must** be identical with the model name plus the word `Manager`. 

Valid manager names with their model name:
```
FooBarManager --> FooBar
PersonManager --> Person
AwesomeMechanicalPieceManager --> AwesomeMechanicalPiece
```
Invalid manager names:
```
FooBar
FooBarmanager
FooBar_Manager
FooManagerBarManager
```

## <a name="methods-list">Properties and methods list</a>
### AbstractModel
- **$id**: ID of the model. The database field must be set to auto increment, as the framework uses this value to check
if the model is already inserted into database or not.
- **insert()**: Create a now row in the table named after the model class by using the model properties as field names.
- **update()**: Same as `insert()` but updates the fields values of the row matching the model ID.
- **delete()**: Delete the table row matching the model ID.
- **getId()**: Returns model's ID.
- **getTableName()**: Returns the model table name by converting the class name to snake_case. This method can be
overridden if a custom table name must be used.

### AbstractManager
- **find(id)**: Returns the model matching the ID from database. If no model matches, an exception is thrown.
- **findWhere(conditions)**: Returns the model filling the conditions. All the conditions are strict equality 
checks (`=` sign). If a different condition check is necessary, feel free to implement a new method.
- **findAll()**: Returns all models managed by this manager.
- **getTableName()**: Returns the model table name by converting the class name (without the word `Manager`) to 
snake_case. This method can be overridden if a custom table name must be used.
- **getModelName(short)**: Returns the model name by trimming the word `Manager` from the class name. If `short` is 
false, will return full class name (package/model). This method can be overridden if a custom manager name must be used.
- **count(column, distinct)**: Counts rows in table using SQL `COUNT`. Default parameters will simply count all rows but
you can specify what column you want to count and if you want it to only count distinct values.

## Changelog
Please take a look at [CHANGELOG.MD](CHANGELOG.MD)

## Please criticize
Please feel free to contact me here or by [email](mailto:gaetanroger.fr@gmail.com) if something bothers you or does not work as expected.

CHANGELOG.MD
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1]
- ADDED - Can now count rows using AbstractManager::count()

#### [1.0.6] - 2017-10-14
- CHANGED - Moved changelog from [README.MD](README.MD) to [CHANGELOG.MD](CHANGELOG.MD)
- CHANGED - Renamed namespaces to match new name
- CHANGED - Wrote better documentation in [README.MD](README.MD) to help new user
install and use the framework

#### [1.0.5] - 2017-10-13
- ADDED - Integrated whole project into Travis CI and Coveralls to get nice badges (green hopefully)
- CHANGED - Renamed project to Minimal PHP Framework
- CHANGED - Rewrote unit tests to use sqlite memory database instead of MySQL

#### [1.0.4] - 2017-07-13
- FIXED - README.MD typo

#### [1.0.3] - 2017-07-13
- ADDED - changelog is now written into README.MD
- FIXED - forgot to bumb version number in v1.0.2

#### [1.0.2] - 2017-07-14
- FIXED - README.MD typo
- FIXED - removed Slim in package description text

#### [1.0.1] - 2017-07-13
- FIXED - removed unused Slim dependency

## [1.0.0] - 2017-07-13
- ADDED - AbstractModel, the parent of all models
- ADDED - AbstractManager, the parent of all managers

test/ModelTest.php
<?php
/**
 * Project base-php-framework
 *
 * @author Gaetan
 * @date   13/07/2017 08:55
 */

namespace Gaetanroger\MinimalPhpFrameworkTest;

use Gaetanroger\MinimalPhpFramework\AbstractModel;
use Gaetanroger\MinimalPhpFrameworkTest\Classes\Dumb\Dumb;
use Gaetanroger\MinimalPhpFrameworkTest\Classes\Dumb\DumbManager;
use PHPUnit\Exception;


class ModelTest extends DatabaseTestBase
{
    public function testGetTableName()
    {
        $myNewModel = $this->getMockForAbstractClass(AbstractModel::class, [], "MyNew");
        $oneAgain = $this->getMockForAbstractClass(AbstractModel::class, [], "OneAgain");
        $dumbModel = new Dumb();
        
        $myNewModelTableName = $myNewModel->getTableName();
        $oneAgainTableName = $oneAgain->getTableName();
        $dumbModelTableName = $dumbModel->getTableName();
        
        $this->assertEquals("my_new", $myNewModelTableName);
        $this->assertEquals("one_again", $oneAgainTableName);
        $this->assertEquals("dumb", $dumbModelTableName);
    }
    
    public function testInsert()
    {
        // Creating true
        $newDumbTrue = new Dumb();
        $newDumbTrue->setName("Dumb name true");
        $newDumbTrue->setCount(42);
        $newDumbTrue->setTimestamp(1499937118);
        $newDumbTrue->setBool(true);
        
        // Creating false
        $newDumbFalse = new Dumb();
        $newDumbFalse->setName("Dumb name false");
        $newDumbFalse->setCount(21);
        $newDumbFalse->setTimestamp(1499937119);
        $newDumbFalse->setBool(false);
        
        // Inserting both
        $newDumbTrue->insert();
        $newDumbFalse->insert();
        
        // Checking ID was set for both
        $this->assertNotNull($newDumbTrue->getId());
        $this->assertNotNull($newDumbFalse->getId());
        $this->assertNotEquals($newDumbTrue->getId(), $newDumbFalse->getId());
        
        /**
         * @var \PDO $database
         */
        $database = $this->getConnection()->getConnection();
        
        
        // Testing true
        
        /**
         * @var Dumb[] $resultsTrue
         */
        $resultsTrue = $database->query(
            'SELECT * FROM ' . $newDumbTrue->getTableName() . ' WHERE name = \'Dumb name true\''
        )->fetchAll(\PDO::FETCH_CLASS, Dumb::class);
        
        $this->assertNotEmpty($resultsTrue);
        $this->assertCount(1, $resultsTrue);
        $this->assertEquals($newDumbTrue->getCount(), $resultsTrue[0]->getCount());
        $this->assertEquals($newDumbTrue->getTimestamp(), $resultsTrue[0]->getTimestamp());
        $this->assertEquals($newDumbTrue->isBool(), $resultsTrue[0]->isBool());
        
        
        // Testing false
        
        
        /**
         * @var Dumb[] $resultsFalse
         */
        $resultsFalse = $database->query(
            'SELECT * FROM ' . $newDumbTrue->getTableName() . ' WHERE name = \'Dumb name false\''
        )->fetchAll(\PDO::FETCH_CLASS, Dumb::class);
        
        $this->assertNotEmpty($resultsFalse);
        $this->assertCount(1, $resultsFalse);
        $this->assertEquals($newDumbFalse->getCount(), $resultsFalse[0]->getCount());
        $this->assertEquals($newDumbFalse->getTimestamp(), $resultsFalse[0]->getTimestamp());
        $this->assertEquals($newDumbFalse->isBool(), $resultsFalse[0]->isBool());
        
        
    }
    
    /**
     * @expectedException \BadMethodCallException
     */
    public function testInsertWithNonNullId()
    {
        $dumb = DumbManager::find(1);
        
        $dumb->insert();
    }
    
    public function testUpdate()
    {
        /**
         * @var Dumb $dumb
         */
        $dumb = DumbManager::find(1);
        
        $dumb->setName('Philip');
        $dumb->setCount(42);
        $dumb->setTimestamp(123456789);
        $dumb->setBool(false);
        $dumb->update();
        
        
        /**
         * @var Dumb $newDumb
         */
        $newDumb = DumbManager::find(1);
        
        $this->assertEquals('Philip', $newDumb->getName());
        $this->assertEquals(42, $newDumb->getCount());
        $this->assertEquals(123456789, $newDumb->getTimestamp());
        $this->assertEquals(false, $newDumb->isBool());
    }
    
    /**
     * @expectedException \UnexpectedValueException
     */
    public function testUpdateNotFound()
    {
        $dumb = new Dumb();
        
        $reflectionClass = new \ReflectionClass($dumb);
        $idProp = $reflectionClass->getProperty('id');
        $idProp->setAccessible(true);
        $idProp->setValue($dumb, 999999);
        
        $dumb->update();
    }
    
    /**
     * @expectedException \Exception
     */
    public function testUpdateWithNullId()
    {
        $dumb = new Dumb();
        $dumb->update();
    }
    
    /**
     * @expectedException Exception
     */
    public function testDelete()
    {
        $dumb = DumbManager::find(1);
        
        $dumb->delete();
        
        $this->assertNull($dumb->getId());
        $this->assertCount(199, DumbManager::findAll());
        DumbManager::find(1);
    }
    
    /**
     * @expectedException \UnexpectedValueException
     */
    public function testDeleteNotFound()
    {
        $dumb = new Dumb();
        
        $reflectionClass = new \ReflectionClass($dumb);
        $idProp = $reflectionClass->getProperty('id');
        $idProp->setAccessible(true);
        $idProp->setValue($dumb, 999999);
        
        $dumb->delete();
    }
    
    /**
     * @expectedException \Exception
     */
    public function testDeleteWithNullId()
    {
        $dumb = new Dumb();
        $dumb->delete();
    }
}


test/UtilsTest.php
<?php
/**
 * Project base-php-framework
 *
 * @author Gaetan
 * @date   13/07/2017 08:20
 */

namespace Gaetanroger\MinimalPhpFrameworkTest;

use Gaetanroger\MinimalPhpFramework\Utils;
use PHPUnit\Framework\TestCase;

class UtilsTest extends TestCase
{
    
    public function testCamelCaseToSnakeCase()
    {
        $camelCase1 = "CamelCase";
        $camelCase2 = "UmlName";
        $camelCase3 = "ILoveLetters";
        $camelCase4 = "OneClassModel";
        
        $snakeCase1 = Utils::camelCaseToSnakeCase($camelCase1);
        $snakeCase2 = Utils::camelCaseToSnakeCase($camelCase2);
        $snakeCase3 = Utils::camelCaseToSnakeCase($camelCase3);
        $snakeCase4 = Utils::camelCaseToSnakeCase($camelCase4);
        
        $this->assertEquals('camel_case', $snakeCase1);
        $this->assertEquals('uml_name', $snakeCase2);
        $this->assertEquals('i_love_letters', $snakeCase3);
        $this->assertEquals('one_class_model', $snakeCase4);
    }
}


test/ManagerTest.php
<?php
/**
 * Project base-php-framework
 *
 * @author Gaetan
 * @date   13/07/2017 14:23
 */

namespace Gaetanroger\MinimalPhpFrameworkTest;

use Gaetanroger\MinimalPhpFramework\AbstractManager;
use Gaetanroger\MinimalPhpFrameworkTest\Classes\Dumb\Dumb;
use Gaetanroger\MinimalPhpFrameworkTest\Classes\Dumb\DumbManager;

class ManagerTest extends DatabaseTestBase
{
    
    public function testGetTableName()
    {
        $myNewManager = $this->getMockForAbstractClass(AbstractManager::class, [], "MyNewManager");
        $oneAgainManager = $this->getMockForAbstractClass(AbstractManager::class, [], "OneAgainManager");
        
        $myNewManagerTableName = $myNewManager::getTableName();
        $oneAgainManagerTableName = $oneAgainManager::getTableName();
        $dumbManagerTableName = DumbManager::getTableName();
        
        $this->assertEquals("my_new", $myNewManagerTableName);
        $this->assertEquals("one_again", $oneAgainManagerTableName);
        $this->assertEquals("dumb", $dumbManagerTableName);
    }
    
    public function testGetModelNameShort()
    {
        $myNewManager = $this->getMockForAbstractClass(AbstractManager::class, [], "MyNewManager");
        $oneAgainManager = $this->getMockForAbstractClass(AbstractManager::class, [], "OneAgainManager");
        
        $myNewManagerModelName = $myNewManager::getModelName();
        $oneAgainManagerModelName = $oneAgainManager::getModelName();
        $dumbManagerModelName = DumbManager::getModelName();
        
        $this->assertEquals("MyNew", $myNewManagerModelName);
        $this->assertEquals("OneAgain", $oneAgainManagerModelName);
        $this->assertEquals("Dumb", $dumbManagerModelName);
    }
    
    public function testGetModelNameLong()
    {
        $modelName = DumbManager::getModelName(false);
        
        $this->assertEquals('Gaetanroger\MinimalPhpFrameworkTest\Classes\Dumb\Dumb', $modelName);
    }
    
    public function testFind()
    {
        $this->getConnection()->getConnection()->query("SELECT * FROM DUMB");
        
        /**
         * @var Dumb $dumb
         */
        $dumb = DumbManager::find(1);
        
        $this->assertNotNull($dumb);
        $this->assertEquals('Aksel', $dumb->getName());
        $this->assertEquals(444, $dumb->getCount());
        $this->assertEquals(1497562479, $dumb->getTimestamp());
        $this->assertEquals(true, $dumb->isBool());
    }
    
    /**
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidFind()
    {
        DumbManager::find(99999);
    }
    
    public function testFindAll()
    {
        $dumbs = DumbManager::findAll();
        
        $this->assertCount(200, $dumbs);
    }
    
    public function testFindWhere()
    {
        /**
         * @var Dumb[] $dumbs
         */
        $dumbs = DumbManager::findWhere(['id' => 1]);
        
        $this->assertCount(1, $dumbs);
        $this->assertEquals('Aksel', $dumbs[0]->getName());
        
        
        $dumbs = DumbManager::findWhere(['count' => 830]);
        
        $this->assertCount(2, $dumbs);
        $this->assertEquals('Annabella', $dumbs[0]->getName());
        $this->assertEquals('Neda', $dumbs[1]->getName());
        
        
        $dumbs = DumbManager::findWhere(['count' => 999999]);
        
        $this->assertCount(0, $dumbs);
        
        
        $dumbs = DumbManager::findWhere(['count' => 18, 'timestamp' => 1494407078]);
        
        $this->assertCount(1, $dumbs);
        $this->assertEquals('Berny', $dumbs[0]->getName());
    }
    
    public function testCountAll()
    {
        $c = DumbManager::count();
        
        $this->assertEquals(200, $c);
    }
    
    public function testCountColumn()
    {
        $cAll = DumbManager::count('bool');
        $cDistinct = DumbManager::count('bool', true);
        
        $this->assertEquals(200, $cAll);
        $this->assertEquals(2, $cDistinct);
    }
    
    /**
     * @expectedException \InvalidArgumentException
     */
    public function testColumnAllDistinctError()
    {
        DumbManager::count('*', true);
    }
}


src/Model.php
<?php


namespace GrBaseFramework;


/**
 * Model class.
 *
 * @author Gaetan
 * @date   12/07/2017
 */
abstract class Model
{
    /**
     * Database instance.
     *
     * @var \PDO $database
     */
    public static $database;
    
    /**
     * Model ID.
     *
     * @var int $id
     */
    protected $id;
    
    /**
     * Return the model unique ID.
     *
     * @return int|null
     */
    public function getId()
    {
        return $this->id;
    }
    
    /**
     * Return model table name.
     *
     * For this to work, the following rules **must** be followed :
     * * A model class name **must** use CamelCase;
     * * A model class name **can** end with `Model`;
     * * A model class name **must** not contain `Model` anywhere else;
     * * The table name **must** use snake_case;
     * * The model class name **must** be the exact conversion from snake_case to CamelCase of the table name.
     *
     * @return string
     */
    public function getTableName(): string
    {
        $reflectionClass = new \ReflectionClass($this);
        $className = $reflectionClass->getShortName();
        $name = str_replace('Model', '', $className);
        
        return Utils::camelCaseToSnakeCase($name);
    }
    
    /**
     * Insert model into database.
     *
     * @return Model
     */
    public function insert(): Model
    {
        $binding = [];
    
        $reflectionClass = new \ReflectionClass($this);
        $properties = $reflectionClass->getProperties(\ReflectionProperty::IS_PROTECTED | \ReflectionProperty::IS_PRIVATE);
        foreach ($properties as $property) {
            $property->setAccessible(true);
            if ($property->getName() === 'id' || $property->getValue($this) === null)
                continue;
        
            $value = $property->getValue($this);
            if ($value === true)
                $value = 1;
            else if ($value === false)
                $value = 0;
        
            $binding[$property->getName()] = $value;
        }
    
        $tableName = $this->getTableName();
        $columns = implode(', ', array_keys($binding));
        $valuesPlaceholders = ':' . implode(', :', array_keys($binding));
    
        $statement = self::$database->prepare("INSERT INTO $tableName ($columns) VALUES ($valuesPlaceholders)");
        $statement->execute($binding);
    
        $this->id = self::$database->lastInsertId();
        
        return $this;
    }
}

src/Manager.php
<?php


namespace GrBaseFramework;


/**
 * Class Manager
 *
 * @author Gaetan
 * @date   13/07/2017
 */
abstract class Manager
{
    /**
     * Database instance.
     *
     * @var \PDO $database
     */
    public static $database;
    
    
    /**
     * Return manager table name.
     *
     * For this to work, the following rules **must** be followed :
     * * A manager class name **must** use CamelCase;
     * * A manager class name **must** end with `Manager`;
     * * A model class name **must** not contain `Manager` anywhere else;
     * * The table name **must** use snake_case;
     * * The manager class name **must** be the exact conversion from snake_case to CamelCase of the table name.
     *
     * @return string
     */
    public static function getTableName(): string
    {
        $reflectionClass = new \ReflectionClass(static::class);
        $className = $reflectionClass->getShortName();
        $name = str_replace('Manager', '', $className);
    
        return Utils::camelCaseToSnakeCase($name);
    }
}

src/AbstractModel.php
<?php


namespace Gaetanroger\MinimalPhpFramework;


/**
 * Model class.
 *
 * A model is a representation of a database entity. Each model is identified by its unsigned integer ID.
 * Please see *README.MD* for naming conventions to follow.
 *
 * @author Gaetan
 * @date   12/07/2017
 */
abstract class AbstractModel
{
    /**
     * Database instance.
     *
     * This \PDO instance must be set before any model is used, as it is the center
     * of any database related operation.
     *
     * @var \PDO $database
     */
    public static $database;
    
    /**
     * Model unique ID.
     *
     * This ID is used as the primary key in all database operations. The model assumes that
     * the ID is in auto increment.
     *
     * @var int $id
     */
    protected $id;
    
    /**
     * Insert model into database.
     *
     * The model assumed the ID field is in auto increment, and will not be set in this method. This method
     * will actually use lastInsertId to set the model ID field to the corresponding value.
     *
     * @return AbstractModel The model itself, to be used in fluent operations.
     * @throws \BadMethodCallException If getId does not return null (meaning the model is already inserted into
     *                                 database).
     */
    public function insert(): AbstractModel
    {
        // If the model is already inserted into database, throwing an exception
        if ($this->getId() !== null) {
            throw new \BadMethodCallException(
                "This model already has an ID (meaning it is already inserted into database). Maybe use update?"
            );
        }
        
        // Name of field associated with value.
        $binding = [];
        
        // Getting properties
        $reflectionClass = new \ReflectionClass($this);
        $properties = $reflectionClass->getProperties(\ReflectionProperty::IS_PROTECTED | \ReflectionProperty::IS_PRIVATE);
        
        
        // For each property of the object
        foreach ($properties as $property) {
            
            // Allowing us to access private values
            $property->setAccessible(true);
            
            // If the value is null, no need to include it in the query (default values are used)
            if ($property->getValue($this) === null) {
                continue;
            }
            
            // Converting boolean to MySQL ones & zeros TINYINT
            $value = $property->getValue($this);
            if ($value === true) {
                $value = 1;
            } else if ($value === false) {
                $value = 0;
            }
            
            // Binding property name to value
            $binding[$property->getName()] = $value;
        }
        
        // Getting table name
        $tableName = $this->getTableName();
        
        // Getting table columns from class attributes
        $columns = implode(', ', array_keys($binding));
        
        // Getting placeholders by adding `:` in front of columns name
        $valuesPlaceholders = ':' . implode(', :', array_keys($binding));
        
        // Running query
        $statement = self::$database->prepare("INSERT INTO $tableName ($columns) VALUES ($valuesPlaceholders)");
        $statement->execute($binding);
        
        // Getting ID
        $this->id = self::$database->lastInsertId();
        
        // Returning $this to allow fluent usage
        return $this;
    }
    
    /**
     * Return the model unique ID.
     *
     * @return int|null
     */
    public function getId()
    {
        return $this->id;
    }
    
    /**
     * Return model table name.
     *
     * For this to work, the following rules **must** be followed :
     * * A model class name **must** use CamelCase;
     * * A model class name **must not** end with `Model`;
     * * The table name **must** use snake_case;
     * * The model class name **must** be the exact conversion from snake_case to CamelCase of the table name.
     *
     * @return string
     */
    public static function getTableName(): string
    {
        // Getting the class name
        $reflectionClass = new \ReflectionClass(static::class);
        $className = $reflectionClass->getShortName();
        
        // Convert to snake_case
        return Utils::camelCaseToSnakeCase($className);
    }
    
    /**
     * Update model in database.
     *
     * The method will use the model ID to update all its other fields.
     *
     * @return AbstractModel The model itself, to be used in fluent operations.
     * @throws \BadMethodCallException If model ID is null (meaning it is not inserted into database).
     */
    public function update(): AbstractModel
    {
        // If the model is not inserted into database, throwing an exception
        if ($this->getId() === null) {
            throw new \BadMethodCallException("Model ID is null (meaning it's not inserted into database). Maybe use insert?");
        }
        
        
        // Getting class properties
        $reflectionClass = new \ReflectionClass($this);
        $properties = $reflectionClass->getProperties(\ReflectionProperty::IS_PROTECTED | \ReflectionProperty::IS_PRIVATE);
        
        // Array of placeholders (columns with `:` before)
        $placeholders = [];
        
        // Name of field associated with value.
        $binding = [];
        
        
        // For each property
        foreach ($properties as $property) {
            
            // If property is the ID, skipping adding it to params
            if ($property->getName() === 'id') {
                continue;
            }
            
            // Allowing us to access private values
            $property->setAccessible(true);
            
            // Values to be places after SET
            $placeholders[] = $property->getName() . ' = :' . $property->getName();
            
            // Binding property name to value
            $binding[':' . $property->getName()] = $property->getValue($this);
        }
        
        
        // Adding the ID to the bindings
        $binding = array_merge($binding, [':id' => $this->getId()]);
        
        // Running query
        $statement = self::$database->prepare(
            'UPDATE ' . $this->getTableName() .
            ' SET ' . implode(', ', $placeholders) .
            ' WHERE id = :id'
        );
        $statement->execute($binding);
        
        // If unexpected rows were modified (or not), throwing an exception
        if ($statement->rowCount() !== 1) {
            throw new \UnexpectedValueException('Row count does not equal one (found ' . $statement->rowCount() . ').');
        }
        
        // Returning $this to allow fluent usage
        return $this;
    }
    
    /**
     * Delete model in database.
     *
     * This method will use the model ID to delete it from database. The model ID will then be set
     * to null.
     *
     * @return AbstractModel The model itself, to be used in fluent operations.
     * @throws \Exception If modified row count does not equal one.
     * @throws \BadMethodCallException If ID is null (meaning that the model is not inserted into database).
     */
    public function delete(): AbstractModel
    {
        // If the model is not inserted into database, throwing an exception
        if ($this->getId() === null) {
            throw new \BadMethodCallException("Model ID is null (meaning it's not inserted into database).");
        }
        
        // Running query
        $statement = self::$database->prepare('DELETE FROM ' . $this->getTableName() . ' WHERE id = :id');
        $statement->bindValue(':id', $this->getId(), \PDO::PARAM_INT);
        $statement->execute();
        
        // If unexpected rows were deleted (or not), throwing an exception
        if ($statement->rowCount() !== 1) {
            throw new \UnexpectedValueException('Row count does not equal one (found ' . $statement->rowCount() . ').');
        }
        
        // Resetting this ID to null
        $this->id = null;
        
        // Returning $this to allow fluent usage
        return $this;
    }
}

src/AbstractManager.php
<?php


namespace Gaetanroger\MinimalPhpFramework;


/**
 * Manager class.
 *
 * A manager defines all static methods not specific to one model.
 * Please see *README.MD* for naming conventions to follow.
 *
 * @author Gaetan
 * @date   13/07/2017
 */
abstract class AbstractManager
{
    /**
     * Database instance.
     *
     * @var \PDO $database
     */
    public static $database;
    
    /**
     * Find the model associated with the given ID.
     *
     * @param   int $id ID of the model to find.
     * @return  AbstractModel
     * @throws  \InvalidArgumentException If nothing was found.
     */
    public static function find(int $id): AbstractModel
    {
        // Finding all models matching the ID (there should be only one)
        $results = static::findWhere(['id' => $id]);
        
        // Throwing an exception if none was found
        if (empty($results)) {
            throw new \InvalidArgumentException("Model not found with id $id.");
        }
        
        // Returning the model
        return $results[0];
    }
    
    /**
     * Find models matching conditions.
     *
     * All the conditions are strict equality checks (`=` sign).
     *
     * @param array $conditions Associative array with key being the column and value being the value to test.
     * @return AbstractModel[] Found models.
     */
    public static function findWhere(array $conditions): array
    {
        // Counting conditions
        $conditionsCount = count($conditions);
        
        // If no conditions given, perform a simple SELECT * on the table
        if ($conditionsCount === 0) {
            return
                static::$database->query('SELECT * FROM ' . static::getTableName())
                    ->fetchAll(\PDO::FETCH_CLASS, self::getModelName(false));
        }
        
        // WHERE conditions
        $where = ' WHERE';
        
        // Array of columns
        $columns = array_keys($conditions);
        
        // Array of values
        $values = array_values($conditions);
        
        
        // Foreach condition
        for ($i = 0; $i < $conditionsCount; ++$i) {
            
            // Appending it to the where string
            $where .= ' ' . $columns[$i] . ' = :' . $columns[$i];
            
            // Adding AND keyword at the end
            if ($i < $conditionsCount - 1) {
                $where .= ' AND';
            }
        }
        
        
        // Preparing query
        $statement = static::$database->prepare('SELECT * FROM ' . static::getTableName() . $where);
        
        // Binding values
        for ($i = 0; $i < $conditionsCount; ++$i) {
            $statement->bindValue(':' . $columns[$i], $values[$i]);
        }
        
        // Running query
        $statement->execute();
        
        return $statement->fetchAll(\PDO::FETCH_CLASS, static::getModelName(false));
    }
    
    /**
     * Return manager table name.
     *
     * For this to work, the following rules **must** be followed :
     * * A manager class name **must** use CamelCase;
     * * A manager class name **must** end with `Manager`;
     * * A manager class name **must** not contain `Manager` anywhere else;
     * * The manager class name **must** be identical with the model name plus the word `Manager`.
     *
     * @return string
     */
    public static function getTableName(): string
    {
        // Getting class name
        $reflectionClass = new \ReflectionClass(static::class);
        $className = $reflectionClass->getShortName();
        
        // Removing `Manager` keyword
        $name = str_replace('Manager', '', $className);
        
        // Converting to snake_case
        return Utils::camelCaseToSnakeCase($name);
    }
    
    /**
     * Return the model name managed by this manager.
     *
     * For this to work, the following rules must be followed:
     * * The model and the manager class **must** be in the same namespace.
     * * The rules specified by `getTableName` in `AbstractManager` and `AbstractModel` **must** both be followed.
     *
     * @param bool $short If true, will only return the short name.
     * @return string
     */
    public static function getModelName(bool $short = true): string
    {
        // Getting class name
        $reflectionClass = new \ReflectionClass(static::class);
        $className = $reflectionClass->getShortName();
        
        // Removing `Manager` keyword
        $shortName = str_replace('Manager', '', $className);
        
        // If short, returning the name
        if ($short) {
            return $shortName;
        }
        
        // Else adding namespace
        return $reflectionClass->getNamespaceName() . '\\' . $shortName;
    }
    
    /**
     * Find all models.
     *
     * @return AbstractModel[]
     */
    public static function findAll(): array
    {
        return static::findWhere([]);
    }
    
    /**
     * Count lines in table using SQL `COUNT`.
     *
     * Default column parameter `*` is used to count all lines without targeting specific column. It is
     * the only value which is not a valid column name that will work. Be careful, using `*` and distinct to true
     * will throw an exception.
     *
     * @param string $column   Column to count.
     * @param bool   $distinct True to count only distinct values when targeting a specific column.
     * @return int Result of count.
     * @throws \InvalidArgumentException if `column` is set to `*` and `distinct` is set to true.
     */
    public static function count(string $column = '*', bool $distinct = false): int
    {
        if ($distinct && $column === '*') {
            throw new \InvalidArgumentException("Cannot use keyword DISTINCT with COUNT(*).");
        }
        
        $tableName = self::getTableName();
        $distinctStr = $distinct ? ' DISTINCT ' : '';
        
        $statement = static::$database->query("SELECT COUNT($distinctStr $column) FROM $tableName");
        
        return $statement->fetch()[0];
    }
}